### PR TITLE
[Snappi] Adding Lossless Response to External pause storm test case

### PR DESCRIPTION
**Pull request by dks0692**
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
Tests the dynamic response of lossless streams when subject to continuous external PFC pause messages intending to fully block their bandwidth. 
<!--

-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

### Approach
#### What is the motivation for this PR?
The motivation here is to check whether DUT’s PFC mechanism responds fast enough to fully stop lossless streams without drops.

#### How did you do it?
It was tested by having back ground and test flow @ 100% line rate on the rx port and rx port sending the PFC pause frames to the tx side thereby pausing the dut to send additional packets

#### How did you verify/test it?
Tested against voq chassis with SONiC installed in it
#### Any platform specific information?

#### Supported testbed topology if it's a new test case?
T2, T1 & T0
### Documentation
Testplan - https://github.com/sonic-net/sonic-mgmt/pull/8373

##Output

AzDevOps@6105354dad99:~/sonic-mgmt/tests$ py.test --inventory ../ansible/snappi-sonic --host-pattern sonic-s6100-dut --testbed vms-snappi-sonic --testbed_file ../ansible/testbed.csv --show-capture=stdout --log-cli-level info --showlocals -ra --allow_recover --skip_sanity --disable_loganalyzer snappi_tests/multidut/pfc/test_lossless_response_to_external_pause_storms.py 
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
============================================================================================================== test session starts ==============================================================================================================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
ansible: 2.8.20
rootdir: /var/AzDevOps/sonic-mgmt/tests, inifile: pytest.ini
plugins: allure-pytest-2.8.22, celery-4.4.7, ansible-2.2.4, forked-1.3.0, html-1.22.1, metadata-1.11.0, repeat-0.9.1, xdist-1.28.0
collecting 1 item                                                                                                                                                                                                                               
-------------------------------------------------------------------------------------------------------------- live log collection --------------------------------------------------------------------------------------------------------------
19:51:59 __init__.pytest_collection_modifyitems   L0582 INFO   | Available basic facts that can be used in conditional skip:
{
  "topo_type": "ptf", 
  "testbed": "vms-snappi-sonic", 
  "topo_name": "ptf64"
}

---------------------------------------------------------------------------------------------------------------- live log setup -----------------------------------------------------------------------------------------------------------------
19:51:59 __init__.set_default                     L0054 INFO   | Completeness level not set during test execution. Setting to default level: CompletenessLevel.basic
19:51:59 __init__.check_test_completeness         L0152 INFO   | Test has no defined levels. Continue without test completeness checks
19:52:06 ptfhost_utils.run_icmp_responder_session L0239 INFO   | Skip running icmp_responder at session level, it is only for dualtor testbed with active-active mux ports.
19:52:06 conftest.creds_on_dut                    L0704 INFO   | dut sonic-s6100-dut belongs to groups [u'snappi-sonic', u'sonic', u'sonic_dell64_40', 'fanout']
19:52:06 conftest.creds_on_dut                    L0728 INFO   | skip empty var file ../ansible/group_vars/all/corefile_uploader.yml
19:52:06 conftest.creds_on_dut                    L0728 INFO   | skip empty var file ../ansible/group_vars/all/env.yml
19:52:06 conftest.nbrhosts                        L0521 INFO   | No VMs exist for this topology: ptf64
19:52:06 conftest.core_dump_and_config_check      L2032 INFO   | Collecting core dumps before test on sonic-s6100-dut
19:52:06 conftest.core_dump_and_config_check      L2041 INFO   | Collecting running config before test on sonic-s6100-dut
19:52:11 __init__.sanity_check                    L0125 INFO   | Skip sanity check according to command line argument
19:52:11 conftest.generate_params_dut_hostname    L1108 INFO   | Using DUTs ['sonic-s6100-dut'] in testbed 'vms-snappi-sonic'
19:52:11 conftest.rand_one_dut_hostname           L0378 INFO   | Randomly select dut sonic-s6100-dut for testing
19:52:11 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture enable_packet_aging_after_test setup starts --------------------
19:52:11 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture enable_packet_aging_after_test setup ends --------------------
19:52:11 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossless_prio setup starts --------------------
19:52:11 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossless_prio setup ends --------------------
19:52:11 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossy_prio setup starts --------------------
19:52:11 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossy_prio setup ends --------------------
19:52:11 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture start_pfcwd_after_test setup starts --------------------
19:52:11 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture start_pfcwd_after_test setup ends --------------------
19:52:11 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_ip setup starts --------------------
19:52:11 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_ip setup ends --------------------
19:52:11 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_port setup starts --------------------
19:52:11 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_port setup ends --------------------
19:52:11 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture snappi_api setup starts --------------------
19:52:11 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture snappi_api setup ends --------------------
19:52:12 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture prio_dscp_map setup starts --------------------
19:52:13 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture prio_dscp_map setup ends --------------------
19:52:13 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture lossless_prio_list setup starts --------------------
19:52:14 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture lossless_prio_list setup ends --------------------
19:52:14 __init__.loganalyzer                     L0045 INFO   | Log analyzer is disabled
19:52:14 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture get_multidut_snappi_ports setup starts --------------------
19:52:14 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture get_multidut_snappi_ports setup ends --------------------
----------------------------------------------------------------------------------------------------------------- live log call -----------------------------------------------------------------------------------------------------------------
19:52:14 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet12 with IP 20.0.1.1/24
19:52:16 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet4 with IP 20.0.2.1/24
19:52:17 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet0 with IP 20.0.3.1/24
19:52:24 connection._warn                         L0246 WARNING| Verification of certificates is disabled
19:52:24 connection._info                         L0243 INFO   | Determining the platform and rest_port using the 10.36.78.135 address...
19:52:24 connection._warn                         L0246 WARNING| Unable to connect to http://10.36.78.135:443.
19:52:24 connection._info                         L0243 INFO   | Connection established to `https://10.36.78.135:443 on linux`
19:52:38 connection._info                         L0243 INFO   | Using IxNetwork api server version 9.30.2212.7
19:52:38 connection._info                         L0243 INFO   | User info IxNetwork/ixnetworkweb/admin-39-18133
19:52:38 snappi_api.info                          L1132 INFO   | snappi-0.9.1
19:52:38 snappi_api.info                          L1132 INFO   | snappi_ixnetwork-0.9.1
19:52:38 snappi_api.info                          L1132 INFO   | ixnetwork_restpy-1.0.64
19:52:39 snappi_api.info                          L1132 INFO   | Config validation 0.012s
19:52:41 snappi_api.info                          L1132 INFO   | Ports configuration 1.723s
19:52:41 snappi_api.info                          L1132 INFO   | Captures configuration 0.163s
19:52:44 snappi_api.info                          L1132 INFO   | Add location hosts [10.36.78.53] 2.348s
19:52:48 snappi_api.info                          L1132 INFO   | Location hosts ready [10.36.78.53] 4.163s
19:52:49 snappi_api.info                          L1132 INFO   | Speed conversion is not require for (port.name, speed) : [('Port 0', u'novusHundredGigNonFanOut'), ('Port 1', u'novusHundredGigNonFanOut'), ('Port 2', u'novusHundredGigNonFanOut')]
19:52:49 snappi_api.info                          L1132 INFO   | Aggregation mode speed change 0.566s
19:52:55 snappi_api.info                          L1132 INFO   | Location preemption [10.36.78.53;6;4, 10.36.78.53;6;2, 10.36.78.53;6;1] 0.110s
19:53:20 snappi_api.info                          L1132 INFO   | Location connect [Port 0, Port 1, Port 2] 25.049s
19:53:21 snappi_api.warning                       L1138 WARNING| Port 0 connectedLinkDown
19:53:21 snappi_api.warning                       L1138 WARNING| Port 1 connectedLinkDown
19:53:21 snappi_api.warning                       L1138 WARNING| Port 2 connectedLinkDown
19:53:21 snappi_api.info                          L1132 INFO   | Location state check [Port 0, Port 1, Port 2] 0.256s
19:53:21 snappi_api.info                          L1132 INFO   | Location configuration 39.486s
19:53:41 snappi_api.info                          L1132 INFO   | Layer1 configuration 20.078s
19:53:41 snappi_api.info                          L1132 INFO   | Lag Configuration 0.087s
19:53:41 snappi_api.info                          L1132 INFO   | Convert device config : 0.243s
19:53:41 snappi_api.info                          L1132 INFO   | Create IxNetwork device config : 0.000s
19:53:42 snappi_api.info                          L1132 INFO   | Push IxNetwork device config : 0.668s
19:53:42 snappi_api.info                          L1132 INFO   | Devices configuration 0.987s
19:53:46 snappi_api.info                          L1132 INFO   | Flows configuration 4.355s
19:53:56 snappi_api.info                          L1132 INFO   | Start interfaces 9.366s
19:53:56 snappi_api.info                          L1132 INFO   | IxNet - The Traffic Item was modified. Please perform a Traffic Generate to update the associated traffic Flow Groups
19:53:56 traffic_generation.run_traffic           L0548 INFO   | Wait for Arp to Resolve ...
19:53:57 traffic_generation.run_traffic           L0565 INFO   | Starting transmit on all flows ...
19:54:01 snappi_api.info                          L1132 INFO   | Flows generate/apply 3.135s
19:54:13 snappi_api.info                          L1132 INFO   | Flows clear statistics 12.200s
19:54:13 snappi_api.info                          L1132 INFO   | Captures start 0.000s
19:54:17 snappi_api.info                          L1132 INFO   | Flows start 3.596s
19:54:17 traffic_generation.run_traffic           L0572 INFO   | Polling DUT for traffic statistics for 20 seconds ...
19:54:37 traffic_generation.run_traffic           L0590 INFO   | DUT polling complete
19:54:53 traffic_generation.run_traffic           L0606 INFO   | All test and background traffic flows stopped
19:54:55 traffic_generation.run_traffic           L0629 INFO   | Dumping per-flow statistics
19:54:56 traffic_generation.run_traffic           L0633 INFO   | Stopping transmit on all remaining flows
19:54:56 traffic_generation.verify_m2o_results    L1045 INFO   | Test Flow 1 -> 0, TX Frames:59865900, RX Frames:59865900
19:54:56 traffic_generation.verify_m2o_results    L1045 INFO   | Test Flow 2 -> 0, TX Frames:59865900, RX Frames:59865900
19:54:56 traffic_generation.verify_m2o_results    L1045 INFO   | Background Flow 1 -> 0, TX Frames:59865900, RX Frames:59865900
19:54:56 traffic_generation.verify_m2o_results    L1045 INFO   | Background Flow 2 -> 0, TX Frames:59865900, RX Frames:59865900
19:54:56 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet12 with ip :20.0.1.1/24
19:54:59 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet4 with ip :20.0.2.1/24
19:55:01 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet0 with ip :20.0.3.1/24
PASSED                                                                                                                                                                                                                                    [100%]
--------------------------------------------------------------------------------------------------------------- live log teardown ---------------------------------------------------------------------------------------------------------------
19:55:03 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture snappi_api teardown starts --------------------
19:55:20 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture snappi_api teardown ends --------------------
19:55:20 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture start_pfcwd_after_test teardown starts --------------------
19:55:21 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture start_pfcwd_after_test teardown ends --------------------
19:55:21 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossy_prio teardown starts --------------------
19:55:21 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossy_prio teardown ends --------------------
19:55:21 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossless_prio teardown starts --------------------
19:55:21 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossless_prio teardown ends --------------------
19:55:21 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture enable_packet_aging_after_test teardown starts --------------------
19:55:23 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture enable_packet_aging_after_test teardown ends --------------------
19:55:23 conftest.core_dump_and_config_check      L2069 INFO   | Collecting core dumps after test on sonic-s6100-dut
19:55:24 conftest.core_dump_and_config_check      L2086 INFO   | Collecting running config after test on sonic-s6100-dut
19:55:25 conftest.core_dump_and_config_check      L2211 WARNING| Core dump or config check failed for test_lossless_response_to_external_pause_storms.py, results: {"core_dump_check": {"new_core_dumps": {"sonic-s6100-dut": []}, "pass": true}, "config_db_check": {"cur_only_config": {"sonic-s6100-dut": {"null": {}}}, "inconsistent_config": {"sonic-s6100-dut": {"null": {}}}, "pre_only_config": {"sonic-s6100-dut": {"null": {"VLAN_MEMBER": {"Vlan2|Ethernet8": {"tagging_mode": "untagged"}, "Vlan2|Ethernet0": {"tagging_mode": "untagged"}, "Vlan2|Ethernet12": {"tagging_mode": "untagged"}, "Vlan2|Ethernet4": {"tagging_mode": "untagged"}}}}}, "pass": false}}


=============================================================================================================== warnings summary ================================================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.plugins.loganalyzer
    self.import_plugin(import_spec)

/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.plugins.sanity_check
    self.import_plugin(import_spec)

/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

snappi_tests/dinesh_response_time_pr/test_lossless_response_to_external_pause_storms.py::test_lossless_response_to_external_pause_storms_test[linecard_configuration_set0-non_chassis_single_line_card]
  /usr/local/lib/python2.7/dist-packages/pytest_ansible/module_dispatcher/v28.py:81: UserWarning: provided hosts list is empty, only localhost is available
    warnings.warn("provided hosts list is empty, only localhost is available")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
------------------------------------------------------------------------------------------------------------ live log sessionfinish -------------------------------------------------------------------------------------------------------------
19:55:25 __init__.pytest_terminal_summary         L0064 INFO   | Can not get Allure report URL. Please check logs
==================== 1 passed, 4 warnings in 207.56 seconds ====================
INFO:root:Can not get Allure report URL. Please check logs

### CODE REVIEW

File: tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+                                                        conn_data,
+                                                        fanout_data,
+                                                        dut_port,
+                                                        pause_prio_list,
+                                                        test_prio_list,
+                                                        bg_prio_list,
+                                                        prio_dscp_map,
+                                                        snappi_extra_params=None):
+    """
+    Run PFC watchdog test in a multi-node (>=3) topoology
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py:47]: is this PFCWD test ?

File: tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+DATA_FLOW_DURATION_SEC = 20
+PAUSE_FLOW_DURATION_SEC = 10
+PAUSE_FLOW_DELAY_SEC = 5
+DATA_FLOW_DELAY_SEC = 0
+SNAPPI_POLL_DELAY_SEC = 2
+TOLERANCE_THRESHOLD = 0.05
+PAUSE_FLOW_RATE = 15
+PAUSE_FLOW_NAME = 'PFC Traffic'
+
+
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py:34]: please add comments for how the test is done and expected results.

File: tests/snappi_tests/multidut/pfc/test_lossless_response_to_external_pause_storms.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                                                         fanout_graph_facts,             # noqa: F811
+                                                         line_card_choice,
+                                                         duthosts,
+                                                         prio_dscp_map,                     # noqa: F811
+                                                         lossless_prio_list,                # noqa: F811
+                                                         linecard_configuration_set,
+                                                         get_multidut_snappi_ports,):       # noqa: F811
+
+    """
+    Run PFC watchdog test under many to one traffic pattern
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/test_lossless_response_to_external_pause_storms.py:32]: please fix the comment

**Comment by vmittal-msft** [comment]
please fix test case failures.

File: tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py (hunk 1 above)
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py:47]: this is a typo error

**Review by vmittal-msft — APPROVED** [review]

**Comment by selldinesh** [comment]
@yxieca @wangxin please help merge.

File: tests/snappi_tests/multidut/pfc/test_lossless_response_to_external_pause_storms.py (hunk 3 above)
Comment (selldinesh) [reply tests/snappi_tests/multidut/pfc/test_lossless_response_to_external_pause_storms.py]: fixed

**Comment by vmittal-msft** [comment]
@developfast  is this good from your side ?

@selldinesh please fix test failures.

**Pull request merged**